### PR TITLE
🧹  split loading container images from opening connections

### PR DESCRIPTION
**Pull request by imilchev**
split loading of containers to a point after connecting. Creating the connection shouldn't be as expensive as it currently is

### CODE REVIEW

File: providers/os/connection/docker_container.go [hunk 1]
```diff
@@ -182,34 +182,40 @@ func NewContainerRegistryImage(id uint32, conf *inventory.Config, asset *invento
...
-			registryOpts = append(registryOpts, remoteOpts[i])
-		}
+		registryOpts = append(registryOpts, remoteOpts...)
 
+		var conn *TarConnection
 		var img v1.Image
-		var rc io.ReadCloser
 		loadedImage := false
 		if asset.Connections[0].Options != nil {
 			if _, ok := asset.Connections[0].Options[COMPRESSED_IMAGE]; ok {
```
Comment (czunker) [inline providers/os/connection/docker_container.go]: I cannot find the place where this is set. Previously, it was set in L212.

**Comment by czunker** [comment]
I got errors when running this with a registry scan.
```
cnspec scan container registry gcr.io/mondoo-dev-262313/ubuntu 
→ loaded configuration from /home/christian/workspace/mondoo/sa-edge-default.json using source --config
→ using service account credentials
→ discover related assets for 1 asset(s)
x tar> could not fetch tar file error="seek /tmp/mondoo.inspection972519126: file already closed"
→ fetch meta information from container registry registry=gcr.io/mondoo-dev-262313/ubuntu
→ resolved image image=gcr.io/mondoo-dev-262313/ubuntu@sha256:e5dd9dbb37df5b731a6688fa49f4003359f6f126958c9c928f937bec69836320 name=gcr.io/mondoo-dev-262313/ubuntu@e5dd9dbb37df
→ resolved image image=gcr.io/mondoo-dev-262313/ubuntu@sha256:a98d9dcf3a34b2b78e78c61d003eb4d7a90d30fd54451686e2f0bd2ef5f602ac name=gcr.io/mondoo-dev-262313/ubuntu@a98d9dcf3a34
→ resolved image image=gcr.io/mondoo-dev-262313/ubuntu@sha256:0925d086715714114c1988f7c947db94064fd385e171a63c07730f1fa014e6f9 name=gcr.io/mondoo-dev-262313/ubuntu@0925d0867157
→ resolved image image=gcr.io/mondoo-dev-262313/ubuntu@sha256:61844ceb1dd55aa110ca578bd4a042200bc64bb5d702c9a19b9fb90409565da0 name=gcr.io/mondoo-dev-262313/ubuntu@61844ceb1dd5
→ synchronize assets





 gcr.io/mondoo-dev-262313/ubuntu@0925d0867157 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────   0%
 gcr.io/mondoo-dev-262313/ubuntu@0925d0867157 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────  n/a score: U
 gcr.io/mondoo-dev-262313/ubuntu@61844ceb1dd5 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────   0%
 gcr.io/mondoo-dev-262313/ubuntu@61844ceb1dd5 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────  n/a score: U
 0/4 scanned 3/4 n/a                          ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────   0%
 0/4 scanned 4/4 n/a                          ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100%
x tar> could not fetch tar file error="seek /tmp/mondoo.inspection854299611: file already closed"
```

**Comment by czunker** [comment]
The same for a single image:
```
cnspec scan container registry gcr.io/mondoo-dev-262313/ubuntu@sha256:e5dd9dbb37df5b731a6688fa49f4003359f6f126958c9c928f937bec69836320 --verbose              INT ✘ │ 05:59:52 
DBG using provider os with connector container
DBG Started a new runtime (2 total)
DBG no need to update provider last-refresh=7m15.141729554s provider=os
DBG Log level set to debug
DBG Started a new runtime (3 total)
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
→ using service account credentials
DBG Started a new runtime (4 total)
DBG initialize client authentication issuer=mondoo/ams kid=//agents.api.mondoo.app/spaces/dazzling-golick-767384/serviceaccounts/2AyMRoi8pfJa6gOqq16QWstmKPK
→ discover related assets for 1 asset(s)
DBG Started a new runtime (5 total)
DBG found valid container registry reference ref=gcr.io/mondoo-dev-262313/ubuntu@sha256:e5dd9dbb37df5b731a6688fa49f4003359f6f126958c9c928f937bec69836320
DBG initialize client authentication issuer=mondoo/ams kid=//agents.api.mondoo.app/spaces/dazzling-golick-767384/serviceaccounts/2AyMRoi8pfJa6gOqq16QWstmKPK
DBG tar> extracted image to temporary file
x tar> could not fetch tar file error="seek /tmp/mondoo.inspection3322352812: file already closed"
```

**Comment by imilchev** [comment]
I made a mistake in the code. I fixed it now and it seems to be working as expected. Also rebased on `main` to get the latest fixes

**Comment by czunker** [comment]
I tried both the registry and a single image. Both worked and also cleaned up the images.

But I'm not sure whether this is working as intended. When scanning a single image, it still removes two temporary files at the end:
```
DBG tar> remove temporary tar file on connection close tar=/tmp/mondoo.inspection1895662558
DBG tar> remove temporary tar file on connection close tar=/tmp/mondoo.inspection1501979699
```
From the PR description, I'd only expected one such message.

File: providers/os/connection/docker_container.go [hunk 2]
```diff
@@ -183,34 +183,40 @@ func NewContainerRegistryImage(id uint32, conf *inventory.Config, asset *invento
...
+				var rc io.ReadCloser
 				// read image from disk
 				img, rc, err = image.LoadImageFromDisk(asset.Connections[0].Options[COMPRESSED_IMAGE])
```
Comment (imilchev) [inline providers/os/connection/docker_container.go]: it's set right here

**Review by czunker — APPROVED** [review]
LGTM. Thanks @imilchev

**Pull request merged**